### PR TITLE
Fix various type-instabilities

### PR DESCRIPTION
**Pull request by sethaxen**
This PR fixes most of the type-instabilities identified by https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/78, the remainder being documented in #328 and https://github.com/JuliaDiff/ChainRulesCore.jl/issues/265 or for type-unstable primals. I'm not yet adding any new tests because that would be redundant with the above PR. Until it is merged, this PR just checks that our current functionality tests still pass.

### CODE REVIEW

File: src/rulesets/LinearAlgebra/blas.jl [hunk 1]
```diff
@@ -17,15 +17,13 @@ rrule(::typeof(BLAS.dot), x, y) = rrule(dot, x, y)
...
     function blas_dot_pullback(ΔΩ)
-        if ΔΩ isa Zero
-            ∂X = Zero()
-            ∂Y = Zero()
-        else
-            ΔΩ = extern(ΔΩ)
-            ∂X = @thunk scal!(n, ΔΩ, blascopy!(n, Y, incy, _zeros(X), incx), incx)
-            ∂Y = @thunk scal!(n, ΔΩ, blascopy!(n, X, incx, _zeros(Y), incy), incy)
-        end
+        ΔΩ = extern(ΔΩ)
```
Comment (oxinabox) [inline src/rulesets/LinearAlgebra/blas.jl:24]: Its not part of this PR But
This line should not have been here.
```suggestion
```

File: src/rulesets/LinearAlgebra/blas.jl [hunk 2]
```diff
@@ -17,15 +17,13 @@ rrule(::typeof(BLAS.dot), x, y) = rrule(dot, x, y)
 
 function rrule(::typeof(BLAS.dot), n, X, incx, Y, incy)
     Ω = BLAS.dot(n, X, incx, Y, incy)
+    function blas_dot_pullback(::Zero)
+        return (NO_FIELDS, DoesNotExist(), Zero(), DoesNotExist(), Zero(), DoesNotExist())
+    end
     function blas_dot_pullback(ΔΩ)
-        if ΔΩ isa Zero
```
Comment (oxinabox) [inline src/rulesets/LinearAlgebra/blas.jl:21]: why is this type unstable?
I mean we don't need this branch because its always applicable to do this.
So its not something to include in the rule.
but i want to understand.

File: src/rulesets/LinearAlgebra/utils.jl [hunk 3]
```diff
@@ -27,3 +27,5 @@ function _eyesubx!(X::AbstractMatrix)
 end
 
 _extract_imag(x) = complex(0, imag(x))
+
+_unionallof(::Type{T}) where {T} = T.name.wrapper
```
Comment (oxinabox) [inline src/rulesets/LinearAlgebra/utils.jl:31]: This name is not obvious to me why we say that.
would
`_unionall_wrapper` be clearear?
Can we also add a docstring?

File: src/rulesets/LinearAlgebra/norm.jl [hunk 4]
```diff
@@ -181,7 +182,11 @@ function rrule(
     return y, norm1_pullback
 end
 
-_norm1_back(x, y, Δy) = sign.(x) .* real(Δy)
+function _norm1_back(x, y, Δy)
+    ∂x = similar(x)
+    ∂x .= sign.(x) .* real(Δy)
+    return ∂x
+end
```
Comment (oxinabox) [inline src/rulesets/LinearAlgebra/norm.jl:189]: It feels bad that we have to do this.
Should we open an issue on JuliaLang/julia about it?

**Review by oxinabox — APPROVED** [review]
Awesome work.
A few points but once those a cleared up, merge when ready.
Thanks

File: src/rulesets/LinearAlgebra/norm.jl (hunk 4 above)
Comment (sethaxen) [reply src/rulesets/LinearAlgebra/norm.jl:189]: No, I think this is necessarily type-unstable, e.g.
```julia
julia> x = Diagonal([1.0, 2.0, 3.0]);

julia> x .* 0
3×3 Diagonal{Float64,Array{Float64,1}}:
 0.0   ⋅    ⋅
  ⋅   0.0   ⋅
  ⋅    ⋅   0.0

julia> x .* 10
3×3 Diagonal{Float64,Array{Float64,1}}:
 10.0    ⋅     ⋅
   ⋅   20.0    ⋅
   ⋅     ⋅   30.0

julia> x .* Inf
3×3 Array{Float64,2}:
  Inf  NaN   NaN
 NaN    Inf  NaN
 NaN   NaN    Inf

julia> x .* NaN
3×3 Array{Float64,2}:
 NaN  NaN  NaN
 NaN  NaN  NaN
 NaN  NaN  NaN
```

File: src/rulesets/LinearAlgebra/utils.jl (hunk 3 above)
Comment (sethaxen) [reply src/rulesets/LinearAlgebra/utils.jl:31]: > would
> `_unionall_wrapper` be clearear?

Yeah naming this is tricky. It's basically `Base.typename` but instead of returning a `TypeName` it returns a `UnionAll`. If `_unionall_wrapper` is clearer, I can do that.

File: src/rulesets/LinearAlgebra/blas.jl (hunk 2 above)
Comment (sethaxen) [reply src/rulesets/LinearAlgebra/blas.jl:21]: Oh, actually, it looks like `extern` caused the type-instability in the type of `InplaceableThunk`. That seems like it should not happen? But I can revert my change and accept the one you provided.

Comment (oxinabox) [reply src/rulesets/LinearAlgebra/blas.jl:21]: Reassigning a variable to contain another type causes type instablilties.
I think it shouldn't (I think the compiler should declare a second slot if that can happen and swap to using that but i guess it is hard to workout in adance, possibly impossible).
This basically makes the slow dynamically types
Often the compilere does compile it away.
But not always.

File: src/rulesets/LinearAlgebra/norm.jl (hunk 4 above)
Comment (mcabbott) [reply src/rulesets/LinearAlgebra/norm.jl:189]: Note that this breaks `gradient(norm, [1,2,3])`.
